<?xml version="1.0" encoding="UTF-8"?>
<launch>
  <!-- Arguments -->
  <arg name="port" default="8888" description="Port to use for TCP server"/>
  <arg name="iphone_name" default="iPhone" description="Connection Name as set on iPhone app"/>
  
  <!-- iPhone Interface Node -->
  <node pkg="iROS" exec="iphone_interface.py" name="iphone_interface" output="screen">
    <param name="port" value="$(var port)"/>
  </node>
  
  <!-- Image Viewers -->
  <!--
  <node pkg="rqt_image_view" exec="rqt_image_view" name="im_viewer_0"/>
  <node pkg="rqt_image_view" exec="rqt_image_view" name="im_viewer_1"/>
  -->
  
  <!-- YOLO Node -->
  <!--
  <node pkg="iROS" exec="yolo_node.py" name="yolo_node" output="screen">
    <param name="iphone_name" value="$(var iphone_name)"/>
  </node>
  -->

  <node pkg="stereo_image_proc" exec="point_cloud_node" name="stereo_image_proc">
    <remap from="left/image_raw" to="/iphone/left/eo"/>
    <remap from="left/camera_info" to="/ihpone/left/camera_info"/>
    <remap from="right/image_raw" to="/iphone/right/eo"/>
    <remap from="right/camera_info" to="/iphone/right/camera_info"/>
  </node>

  <!-- Rectify Left EO (Camera) -->
  <node pkg="image_proc" exec="rectify_node" name="rectify_left_eo" output="screen">
    <remap from="image" to="/iphone/left/eo/raw"/>
    <remap from="image_rect" to="/iphone/left/eo/rect"/>
  </node>

  <!-- Rectify Left depth -->
  <node pkg="image_proc" exec="rectify_node" name="rectify_left_depth" output="screen">
    <remap from="image" to="/iphone/left/depth/raw"/>
    <remap from="image_rect" to="/iphone/left/depth/rect"/>
  </node>

  <!-- Make decimated eo image stream -->
  <node pkg="image_proc" exec="resize_node" name="resize_left_eo" output="screen">
    <remap from="image/image_raw" to="/iphone/left/eo/rect"/>
    <remap from="image/camera_info" to="/iphone/left/eo/camera_info"/>
    <remap from="resize/image_raw" to="/iphone/left/eo/resized/image"/>
    <remap from="resize/camera_info" to="/iphone/left/eo/resized/camera_info"/>
    <param name="use_scale" value="false"/>
    <param name="width" value="256"/>
    <param name="height" value="192"/>
    <param name="interpolation" value="1"/>
  </node>

  <!-- Make pointcloud -->
  <node pkg="depth_image_proc" exec="point_cloud_xyzrgb_node" name="left_pc" output="screen">
    <remap from="depth_registered/image_rect" to="/iphone/left/depth/rect"/>
    <remap from="rgb/image_rect_color" to="/iphone/left/eo/resized/image"/>
    <remap from="rgb/camera_info" to="/iphone/left/eo/resized/camera_info"/>
    <remap from="points" to="/iphone/left/points"/>
  </node>

  <!-- Static transform publisher -->
  <node pkg="tf2_ros" exec="static_transform_publisher" name="static_tf_pub_map_to_world"
        args="--x 0 --y 0 --z 0 --roll 0 --pitch 0 --yaw 0 --frame-id map --child-frame-id camera_depth_optical_frame"/>
  
  <!-- Launch RViz -->
  <node pkg="rviz2" exec="rviz2" name="rviz2" output="screen"
    args="-d $(find-pkg-share iROS)/rviz/config.rviz"/>
</launch>
